File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {

    public int solution(int[][] maps) {
        final int n = maps.length;
        final int m = maps[0].length;
        final Queue<Position> que = new LinkedList<>();
        final boolean[][] visited = new boolean[n][m];

        final int[] dx = new int[]{1, -1, 0, 0};
        final int[] dy = new int[]{0, 0, 1, -1};

        // 시작 위치 처리
        visited[0][0] = true;
        que.add(new Position(0, 0, 1));

        int answer = Integer.MAX_VALUE;

        while (!que.isEmpty()) {
            final Position pos = que.poll();

            if (pos.x == n - 1 && pos.y == m - 1) {
                answer = Math.min(answer, pos.cnt);
            }
            
            for (int i = 0; i < 4; i++) {
                int nx = pos.x + dx[i];
                int ny = pos.y + dy[i];

                if (nx > n - 1 || nx < 0 || ny > m - 1 || ny < 0) {
                    continue;
                }

                if (!visited[nx][ny] && maps[nx][ny] == 1) {
                    visited[nx][ny] = true;
                    que.add(new Position(nx, ny, pos.cnt + 1));
                }
            }
        }
        return answer == Integer.MAX_VALUE ? -1 : answer;
    }

    private static class Position {
        private final int x;
        private final int y;
        private final int cnt;

        public Position(final int x, final int y, final int cnt) {
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
"""

"""
from collections import deque

def solution(maps):
    n, m = len(maps), len(maps[0])
    INF = float('inf')
    answer = INF
    
    def dfs(i, j):
        nonlocal answer
        que = deque()
        que.appendleft((i, j, 1))
        visited = [[False for _ in range(m)] for _ in range(n)]
        visited[i][j] = True
        
        while que:
            x, y, steps = que.pop()
            
            if x == n-1 and y == m-1:
                answer = min(answer, steps)
                
            for dir_x, dir_y in [(-1,0), (1,0), (0,-1), (0,1)]:
                pos_x = x + dir_x
                pos_y = y + dir_y
            
                if pos_x < 0 or pos_x >= n or pos_y < 0 or pos_y >= m:
                    continue
                
                if maps[pos_x][pos_y] == 0 or visited[pos_x][pos_y]:
                    continue
                
                visited[pos_x][pos_y] = True
                que.appendleft((pos_x, pos_y, steps + 1))
                
    dfs(0, 0)
    return -1 if answer == INF else answer
